internal/libs/reservoir/reservoir.go

```diff
@@ -92,26 +91,42 @@ func (s *Sampler[T]) Percentile(p float64) (T, bool) {
 		return zero, false
 	}
 
-	p = min(max(p, 0.0), 1.0) // Clamp p to [0.0, 1.0].
+	// If we have a cached value and p is unchanged:
+	cachePresent := !s.lastCalc.IsZero()
+	if cachePresent && p == s.lastP {
```

Comment:
nit: the performance profile of this data structure heavily depends on p being fixed. I'd recommend making p a constructor parameter instead.

Reply:
OR instead of keeping lastVal, you can just cache the sorted slice instead (a bit more complicated, I suppose)